File: pythonlib/src/cealn/exec.py
from __future__ import annotations

from typing import Dict, List, Optional

from .provider import Provider, Field
from .label import Label


class Executable(Provider):
    name = Field(Optional[str])

    context = Field(Label)
    executable_path = Field(str)

    search_paths = Field(List[str], default=[])
    library_search_paths = Field(List[str], default=[])

    def add_dependency_executable(self, rule, *executables, context_id=None) -> Executable:
        search_paths = [*self.search_paths]
        library_search_paths = [*self.library_search_paths]
        new_context = rule.new_depmap(id=context_id)
        if self.context:
            new_context.merge(self.context)
        for executable in executables:
            new_context.merge(executable.context)
            search_paths += executable.search_paths
            library_search_paths += executable.library_search_paths
        new_context = new_context.build()

        return Executable(
            context=new_context,
            executable_path=self.executable_path,
            search_paths=search_paths,
            library_search_paths=library_search_paths,
        )


class LinuxExecutePlatform(Provider):
    execution_sysroot = Field(Label)
    execution_sysroot_input_dest = Field(str)
    execution_sysroot_output_dest = Field(str)
    execution_sysroot_exec_context_dest = Field(str)

    uid = Field(int)
    gid = Field(int)

    standard_environment_variables = Field(Dict[str, str])

    use_fuse = Field(bool, default=True)
    use_interceptor = Field(bool, default=True)

    def substitute(self, value: str):
        return value.replace("%[srcdir]", self.execution_sysroot_input_dest).replace(
            "%[execdir]", self.execution_sysroot_exec_context_dest
        )


class MacOSExecutePlatform(Provider):
    execution_sysroot_extra = Field(Label)

File: crates/data/src/action/exec.rs
use std::collections::BTreeMap;

use cealn_data_derive_provider_serde::ProviderSerde;
use serde::{Deserialize, Serialize};

use crate::{depmap::DepmapType, Label};

#[derive(Clone, PartialEq, Eq, Hash, Debug, ProviderSerde)]
pub struct Executable<I: DepmapType> {
    pub name: Option<String>,
    pub executable_path: String,
    pub context: Option<I::DepmapReference>,
    pub search_paths: Vec<String>,
    pub library_search_paths: Vec<String>,
}

/// Holds platform-specific execution parameters
#[derive(Clone, PartialEq, Eq, Hash, Debug)]
pub enum ExecutePlatform<I: DepmapType> {
    Linux(LinuxExecutePlatform<I>),
    MacOS(MacOSExecutePlatform<I>),
}

/// Linux-specific execution parameters
#[derive(Clone, PartialEq, Eq, Hash, Debug, ProviderSerde)]
pub struct LinuxExecutePlatform<I: DepmapType> {
    /// The static files that make up the root filesystem in which the program will be executed
    pub execution_sysroot: I::DepmapReference,

    pub execution_sysroot_input_dest: String,
    pub execution_sysroot_output_dest: String,
    pub execution_sysroot_exec_context_dest: String,

    pub uid: u32,
    pub gid: u32,

    pub standard_environment_variables: BTreeMap<String, String>,

    pub use_fuse: bool,
    pub use_interceptor: bool,
}

#[derive(Clone, PartialEq, Eq, Hash, Debug, ProviderSerde)]
pub struct MacOSExecutePlatform<I: DepmapType> {
    pub execution_sysroot_extra: I::DepmapReference,
}

impl<I: DepmapType> Serialize for ExecutePlatform<I> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            ExecutePlatform::Linux(linux) => linux.serialize(serializer),
            ExecutePlatform::MacOS(macos) => macos.serialize(serializer),
        }
    }
}

impl<'de, I: DepmapType> Deserialize<'de> for ExecutePlatform<I> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        // FIXME: detect
        let linux = LinuxExecutePlatform::<I>::deserialize(deserializer)?;
        Ok(ExecutePlatform::Linux(linux))
    }
}
